flow360/component/geometry.py

```diff
@@ -186,28 +135,84 @@ def submit(self, progress_callback=None) -> Geometry:
 
         if not shared_account_confirm_proceed():
             raise Flow360ValueError("User aborted resource submit.")
+        # The first geometry is assumed to be the main one.
+        req = NewGeometryRequest(
+            name=self.name,
+            solver_version=solver_version,
+            tags=self.tags,
+            files=[
+                GeometryFileMeta(
+                    name=os.path.basename(file_path),
+                    type="main" if item_index == 0 else "dependency",
+                )
+                for item_index, file_path in enumerate(self.file_names)
+            ],
+            # pylint: disable=fixme
+            # TODO: remove hardcoding
+            parent_folder_id="ROOT.FLOW360",
+            length_unit=self.length_unit,
+            description=description,
+        )
 
-        data = {
-            "geometryName": self.name,
-            "tags": self.tags,
-        }
-
-        if self.solver_version:
-            data["solverVersion"] = self.solver_version
-
-        resp = RestApi(GeometryInterface.endpoint).post(data)
+        ##:: Create new Geometry resource and project
+        resp = RestApi(GeometryInterface.endpoint).post(req.model_dump(by_alias=True))
         info = GeometryMeta(**resp)
-        self._id = info.id
-        submitted_geometry = Geometry(self.id)
 
-        remote_file_names = []
-        for geometry_file in self.file_names:
-            remote_file_name = os.path.basename(geometry_file)
-            file_name_to_upload = geometry_file
-            submitted_geometry._upload_file(
-                remote_file_name, file_name_to_upload, progress_callback=progress_callback
+        ##:: upload geometry files
+        geometry = Geometry(info.id)
+        heartbeat_info = {"resourceId": info.id, "resourceType": "Geometry", "stop": False}
+        heartbeat_thread = threading.Thread(target=_post_upload_heartbeat, args=(heartbeat_info,))
+        heartbeat_thread.start()
+        for file_path in self.file_names:
+            geometry._web._upload_file(
+                remote_file_name=os.path.basename(file_path),
+                file_name=file_path,
+                progress_callback=progress_callback,
             )
-            remote_file_names.append(remote_file_name)
-        submitted_geometry._complete_upload(remote_file_names)
-        log.info(f"Geometry successfully submitted: {submitted_geometry.short_description()}")
-        return submitted_geometry
+        heartbeat_info["stop"] = True
+        heartbeat_thread.join()
+        ##:: kick off pipeline
+        RestApi(GeometryInterface.endpoint, id=geometry._web.id).patch(
+            {"action": "Success"}, method="files"
+        )
```

Comment:
```suggestion
        geometry._web.patch(
            {"action": "Success"}, method="files"
        )
```

Reply:
also, before it was called from from function `_complete_upload` which was more informative. This "patch" is a bit enigmatic.

Reply:
I see. So direct RestAPI calls should be avoided because it is a too low level call? All/most communication should be initiated with the `_web` member?

Reply:
yes, exactly. RestApi is a bit too low level, sometimes hard to avoid.